File: golang/remove_nth_node_from_end_of_list.go
package main

import (
	"golang/model"
)

/**
Space	: O(1)
Time	: O(n)
*/

func removeNthFromEnd(head *model.ListNode, n int) *model.ListNode {
	p := head
	l := 0

	for p != nil {
		l++
		p = p.Next
	}

	if n == l {
		return head.Next
	}

	p = head
	stop := l - n
	for stop-1 > 0 {
		stop--
		p = p.Next
	}

	if p.Next.Next == nil {
		p.Next = nil
	} else {
		p.Next = p.Next.Next
	}

	return head
}

File: python3/remove_nth_node_from_end_of_list.py
from .model import ListNode

"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        fast, slow = head, head

        for _ in range(n):
            fast = fast.next

        if not fast:
            return head.next

        while fast.next:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next

        return head
